Common++/header/MacAddress.h

```diff
@@ -29,60 +29,60 @@ namespace pcpp
 	public:
 		/**
 		 * Default constructor for this class.
-		 * Initializes object to me MacAddress::Zero
+		 * Initializes the address as 00:00:00:00:00:00.
 		 */
-		MacAddress() : m_IsValid(true) { memset(m_Address, 0, sizeof(m_Address)); }
+		MacAddress() {};
 
 		/**
-		 * A constructor that creates an instance of the class out of a byte array. The byte array length must be equal or greater to 6
-		 * (as MAC address is 6-byte long)
-		 * @todo there is no verification array length >= 6. If this is not the case, address will read uninitialized memory
+		 * A constructor that creates an instance of the class out of a byte array. The byte array length 6 (as MAC address is 6-byte long),
+		 * and the remaining bytes are ignored. If the byte array is invalid, the constructor throws.
 		 * @param[in] addr A pointer to the byte array containing 6 bytes representing the MAC address
 		 */
-		MacAddress(const uint8_t* addr) : m_IsValid(true) { memcpy(m_Address, addr, sizeof(m_Address)); }
-
-		/**
-		 * A constructor that creates an instance of the class out of a (char*) string.
-		 * If the string doesn't represent a valid MAC address, instance will be invalid, meaning isValid() will return false
-		 * @param[in] addr A pointer to the (char*) string
-		 */
-		MacAddress(const char* addr) { init(addr); }
+		explicit MacAddress(const uint8_t* addr) { memcpy(m_Address, addr, sizeof(m_Address)); }
```

Comment:
@seladb I intentionally make it `explicit`. I believe employing implicit casting, as demonstrated in`layer->setTargetAddr("00:90:27:85:cf:01");,` is not a recommended practice. In my opinion, we should address this concern; otherwise, individuals might mistakenly assume that `setTargetAddr` accepts a string as a parameter initially. Moreover, the PR is already a **breaking change** regardless.

Reply:
I agree this PR already contains breaking changes, but I believe this change is not required. Why do you think implicit casting isn't recommended in this case?

Reply:
You're right, maybe that's the reason I kept the `char*` constructor and didn't remove it a long time ago... 🤔

Reply:
@seladb  Certainly, and we've previously talked about the reasons for rejecting the `char*` constructor. I think prioritizing a clear interface design is more valuable than relying on the convenience of implicit casting.

Reply:
I'm curious to hear what @egecetin and @clementperon think about it. This PR introduces breaking changes anyway, but this one might be a bit more painful than the others. On the other hand, I agree with @tigercosmos that we should have a more clear interface

Reply:
Implicit casting is not recommended because it can lead bugs because of it _might_ cast from another type accidentally if the developer misunderstand something. I agree in this situation it gives the flexibility to developer and we should prevent too many changes somehow but to be honest I'm not sure about should we add explicit or not. Maybe with this PR we can only make other changes (without adding `explicit` keyword and related changes), then introduce a `Fix explicit issues` PR later?

I remember the same issue from one of my previous PR, reverted explicit from some class constructors because of this breaking change. I can't remember which one exactly (probably one of pre-commit or maybe cppcheck related PRs) but added this ignore rule because of them,

https://github.com/seladb/PcapPlusPlus/blob/e5bcc908e921ce336c621f721a6ed3ecbbe45b06/cppcheckSuppressions.txt#L9-L13

If we fix the issue, we can also remove this ignore rules.

Reply:
Avoiding cppcheck checks appears to resemble an anti-pattern. In addition, `explicit` is the cause of many lines of changes, the main reason is the char* constructor.

Reply:
@seladb @egecetin  well, I think if we want to have `explicit`, it's the best chance now. However, if removing `explicit` can make this PR merged, I can compromise.

Reply:
I tried throwing this problem at ChatGPT and it offered the following solution that seems to work to convert a string to `MacAddress` explicitly`:

```cpp
template<typename T>
MacAddress(const T& addr, typename std::enable_if<std::is_convertible<T, std::string>::value>::type* = nullptr)
{
    ...
}
```

Do you think it might work? 🤔

Reply:
@seladb  smart! just added a template constructor. It's clearer now.

```
explicit MacAddress(const std::string& addr);

template<typename T, typename = typename std::enable_if<std::is_convertible<T, std::string>::value>::type>
MacAddress(const T& addr) : MacAddress(static_cast<std::string>(addr)) {};
```

Reply:
@clementperon  we already disable the cppcheck for implicit casting. We were discussing whether to keep `explicit` or not. Now I think we have the answer, we can still keep the explicit constructor for std::string, but at the same time, have a template constructor (for std::string convertible string) do implicit casting.

Reply:
Then what's the difference between an implicit constructor vs explicit constructor + implicit cast ?